File: packages/sdk/src/client/facilitator.ts
import {
  FacilitatorNotConfiguredError,
  FacilitatorRejectedError,
  MissingTokenDomainSeparatorError,
  OutcomeUnknownError,
  Permit2AllowanceRequiredError,
  SponsorshipError,
  SponsorshipTransportError,
} from "@/errors";
import type { FetchFn } from "@/rpc";
import { isRecord } from "@/serde";

export const NAMES_THE_REQUEST: ReadonlySet<string> = new Set([
  "INVALID_REQUEST",
  "MALFORMED_SIGNATURE",
  "SIGNATURE_MISMATCH",
  "EXPIRED",
  "NOT_YET_VALID",
  "NONCE_ALREADY_USED",
  "SIMULATION_REVERTED",
]);

export const NAMES_THE_CLAIM: ReadonlySet<string> = new Set([
  "INVALID_REQUEST",
  "ACTION_UNAVAILABLE",
  "ACTION_MISMATCH",
  "SIMULATION_REVERTED",
  "REVERTED_ON_CHAIN",
  "RECEIPT_UNAVAILABLE",
]);

export const NAMES_THE_PAYMENT: ReadonlySet<string> = new Set([
  ...NAMES_THE_CLAIM,
  "MALFORMED_SIGNATURE",
  "SIGNATURE_MISMATCH",
  "EXPIRED",
  "NOT_YET_VALID",
  "NONCE_ALREADY_USED",
  "INSUFFICIENT_BALANCE",
]);

const NEVER_ARRIVED_CODES = new Set([
  "ECONNREFUSED",
  "ENOTFOUND",
  "EAI_AGAIN",
  "EHOSTUNREACH",
  "ENETUNREACH",
  "EADDRNOTAVAIL",
  "ERR_INVALID_URL",
  "UND_ERR_CONNECT_TIMEOUT",
]);

export function neverArrived(err: unknown): boolean {
  for (
    let cur: unknown = err;
    cur instanceof Error;
    cur = (cur as { cause?: unknown }).cause
  ) {
    const code = (cur as { code?: unknown }).code;
    if (typeof code !== "string") continue;
    if (NEVER_ARRIVED_CODES.has(code) || code.startsWith("ERR_TLS_")) {
      return true;
    }
  }
  return false;
}

export class Facilitator {
  private baseUrl?: string;
  private fetchFn: FetchFn;

  constructor(baseUrl: string | undefined, fetchFn: FetchFn = fetch) {
    this.baseUrl = baseUrl;
    this.fetchFn = fetchFn;
  }

  isConfigured(): boolean {
    return this.baseUrl !== undefined;
  }

  async aclose(): Promise<void> {
    // no-op for symmetry with the Python SDK
  }

  async post(
    path: string,
    body: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    if (this.baseUrl === undefined) {
      throw new FacilitatorNotConfiguredError();
    }
    const base = this.baseUrl.endsWith("/") ? this.baseUrl : `${this.baseUrl}/`;
    const fetchFn = this.fetchFn;
    let response: Response;
    try {
      response = await fetchFn(base + path, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify(body),
      });
    } catch (err) {
      const message = `facilitator request failed: ${err instanceof Error ? err.message : String(err)}`;
      throw neverArrived(err)
        ? new SponsorshipTransportError(message)
        : new OutcomeUnknownError(message);
    }

    if (!response.ok) {
      const text = await response.text().catch(() => "");
      const message = `facilitator returned ${response.status}: ${text}`;
      if (
        response.status >= 400 &&
        response.status < 500 &&
        response.status !== 408
      ) {
        throw new SponsorshipTransportError(message);
      }
      throw new OutcomeUnknownError(message);
    }

    let payload: unknown;
    try {
      payload = await response.json();
    } catch (err) {
      throw new OutcomeUnknownError(
        `malformed facilitator response: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
    if (!isRecord(payload)) {
      throw new OutcomeUnknownError(
        "malformed facilitator response: not an object",
      );
    }
    return payload;
  }
}

export function confirmFacilitatorEcho(
  field: string,
  raw: unknown,
  expected: string,
): string {
  if (raw === null || raw === undefined) {
    return expected;
  }
  if (String(raw).toLowerCase() === expected.toLowerCase()) {
    return expected;
  }
  throw new OutcomeUnknownError(
    `facilitator echoed ${field} ${raw}, expected ${expected}`,
  );
}

export function eip2612NonceFrom(
  payload: Record<string, unknown>,
): bigint | undefined {
  const allowance = payload.permit2Allowance;
  if (!isRecord(allowance)) {
    return undefined;
  }
  const raw = allowance.eip2612Nonce;
  if (raw === null || raw === undefined) {
    return undefined;
  }
  try {
    return BigInt(String(raw));
  } catch {
    return undefined;
  }
}

export function rejectionError(
  payload: Record<string, unknown>,
  message: unknown,
): SponsorshipError {
  const code = payload.errorCode || "UNKNOWN";
  const text = String(message || payload.error || "facilitator gave no reason");
  if (code === "PERMIT2_ALLOWANCE_REQUIRED") {
    return new Permit2AllowanceRequiredError(text, eip2612NonceFrom(payload));
  }
  if (code === "NO_RELAYER_CONFIGURED" || code === "NO_RELAYER") {
    return new FacilitatorNotConfiguredError();
  }
  return new FacilitatorRejectedError(
    String(code),
    text,
    Boolean(payload.retryable),
  );
}

export function sponsorshipUnavailable(
  err: unknown,
  namesTheRequest: ReadonlySet<string>,
): boolean {
  if (err instanceof OutcomeUnknownError) {
    return false;
  }
  if (err instanceof Permit2AllowanceRequiredError) {
    return false;
  }
  if (err instanceof FacilitatorRejectedError) {
    return !namesTheRequest.has(err.code);
  }
  return err instanceof SponsorshipError;
}

export function refusesTheAuthorization(err: unknown): boolean {
  if (err instanceof MissingTokenDomainSeparatorError) {
    return true;
  }
  return (
    err instanceof FacilitatorRejectedError &&
    (err.code === "SIMULATION_REVERTED" ||
      err.code === "UNSUPPORTED_TRANSFER_METHOD")
  );
}

File: packages/sdk-python/fourmica_sdk/client/facilitator.py
"""Transport for the service that submits signed authorizations and pays the
gas for them. Port of ``sdk-rust/src/client/facilitator.rs``.

The facilitator reports rejections in the body with a 200, so a non-success
status is a transport or routing problem rather than a refused request; a
request that provably never arrived is a :class:`SponsorshipTransportError`,
while anything that may have been acted on is an :class:`OutcomeUnknownError`
— retrying those blindly risks paying twice.
"""

from __future__ import annotations

from typing import Any, Dict, Optional

import httpx

from ..errors import (
    FacilitatorNotConfiguredError,
    FacilitatorRejectedError,
    FourMicaError,
    OutcomeUnknownError,
    Permit2AllowanceRequiredError,
    SponsorshipError,
    SponsorshipTransportError,
)

# Rejections that describe the request itself rather than the facilitator's
# willingness to pay. The caller's own transaction would fail for the same
# reason, so falling back to self-funding just burns their gas on a revert.
NAMES_THE_REQUEST = frozenset(
    {
        "INVALID_REQUEST",
        "MALFORMED_SIGNATURE",
        "SIGNATURE_MISMATCH",
        "EXPIRED",
        "NOT_YET_VALID",
        "NONCE_ALREADY_USED",
        "SIMULATION_REVERTED",
    }
)

# Additional claim-shaped rejections: the self-funded path resolves the same
# terms from the same core and submits to the same contract.
NAMES_THE_CLAIM = frozenset(
    {
        "INVALID_REQUEST",
        "ACTION_UNAVAILABLE",
        "ACTION_MISMATCH",
        "SIMULATION_REVERTED",
        "REVERTED_ON_CHAIN",
        "RECEIPT_UNAVAILABLE",
    }
)

# Beyond the claim codes, the debtor's side of the bargain: a refused
# signature means this SDK signed over the wrong terms, and an insufficient
# balance fails the self-funded route just the same.
NAMES_THE_PAYMENT = NAMES_THE_CLAIM | frozenset(
    {
        "MALFORMED_SIGNATURE",
        "SIGNATURE_MISMATCH",
        "EXPIRED",
        "NOT_YET_VALID",
        "NONCE_ALREADY_USED",
        "INSUFFICIENT_BALANCE",
    }
)


class Facilitator:
    def __init__(
        self,
        base_url: Optional[str],
        client: Optional[httpx.AsyncClient] = None,
    ) -> None:
        # None when none was configured; every call then fails with
        # FacilitatorNotConfiguredError rather than silently doing nothing.
        self._base_url = base_url
        self._client = client or httpx.AsyncClient(timeout=30.0)

    def is_configured(self) -> bool:
        return self._base_url is not None

    async def aclose(self) -> None:
        await self._client.aclose()

    async def post(self, path: str, body: Dict[str, Any]) -> Dict[str, Any]:
        if self._base_url is None:
            raise FacilitatorNotConfiguredError()
        base = self._base_url if self._base_url.endswith("/") else self._base_url + "/"

        try:
            response = await self._client.post(base + path, json=body)
        except (httpx.ConnectError, httpx.InvalidURL, httpx.UnsupportedProtocol) as exc:
            raise SponsorshipTransportError(
                f"facilitator request failed: {exc}"
            ) from exc
        except httpx.HTTPError as exc:
            # The request may have gone out; whether it was acted on is unknown.
            raise OutcomeUnknownError(f"facilitator request failed: {exc}") from exc

        if not response.is_success:
            message = f"facilitator returned {response.status_code}: {response.text}"
            if 400 <= response.status_code < 500 and response.status_code != 408:
                raise SponsorshipTransportError(message)
            raise OutcomeUnknownError(message)

        try:
            payload = response.json()
        except Exception as exc:
            raise OutcomeUnknownError(f"malformed facilitator response: {exc}") from exc
        if not isinstance(payload, dict):
            raise OutcomeUnknownError("malformed facilitator response: not an object")
        return payload


def confirm_facilitator_echo(field: str, raw: Optional[str], expected: str) -> str:
    """Checks a value the facilitator echoed back against what was asked for,
    taking the request's own value when the echo is omitted. One that
    disagrees — or cannot be read — means the receipt would describe a
    transaction nobody asked for, and is refused as an unknown outcome."""
    if raw is None:
        return expected
    if str(raw).lower() == str(expected).lower():
        return expected
    raise OutcomeUnknownError(f"facilitator echoed {field} {raw}, expected {expected}")


def eip2612_nonce(payload: Dict[str, Any]) -> Optional[int]:
    """The EIP-2612 nonce attached to a ``PERMIT2_ALLOWANCE_REQUIRED``
    rejection — the one value a client with no chain access cannot compute."""
    allowance = payload.get("permit2Allowance")
    if not isinstance(allowance, dict):
        return None
    raw = allowance.get("eip2612Nonce")
    if raw is None:
        return None
    try:
        return int(str(raw), 0)
    except ValueError:
        return None


def rejection_error(
    payload: Dict[str, Any], message: Optional[str]
) -> SponsorshipError:
    """The typed rejection for a ``success: false`` / ``isValid: false`` body.
    ``errorCode`` is carried verbatim so a caller can branch on a code this
    SDK predates; absent ``retryable`` means "not retryable" — a facilitator
    that omits it is not promising anything."""
    code = payload.get("errorCode") or "UNKNOWN"
    text = message or payload.get("error") or "facilitator gave no reason"
    if code == "PERMIT2_ALLOWANCE_REQUIRED":
        return Permit2AllowanceRequiredError(text, eip2612_nonce(payload))
    if code in ("NO_RELAYER_CONFIGURED", "NO_RELAYER"):
        return FacilitatorNotConfiguredError()
    return FacilitatorRejectedError(
        code=str(code), message=str(text), retryable=bool(payload.get("retryable"))
    )


def sponsorship_unavailable(exc: FourMicaError, names_the_request: frozenset) -> bool:
    """Whether an error means "nobody sponsored this", as opposed to "this
    request is bad" or "we do not know what happened". Only the first is worth
    paying for a self-funded retry: a rejection naming the request would
    revert the caller's own transaction too, and an unknown outcome may mean
    the facilitator already submitted."""
    if isinstance(exc, OutcomeUnknownError):
        return False
    if isinstance(exc, Permit2AllowanceRequiredError):
        return False
    if isinstance(exc, FacilitatorRejectedError):
        return exc.code not in names_the_request
    return isinstance(exc, SponsorshipError)


def refuses_the_authorization(exc: FourMicaError) -> bool:
    """Whether a rejection means "this token cannot take an EIP-3009
    authorization" rather than "this request is bad". A token without
    ``receiveWithAuthorization`` reverts opaquely, reported as a failed
    simulation — retrying over Permit2 is a guess, but a cheap one: the
    simulation spent no gas, and a genuinely bad request fails the second
    route with its own error."""
    from ..errors import MissingTokenDomainSeparatorError

    if isinstance(exc, MissingTokenDomainSeparatorError):
        return True
    return isinstance(exc, FacilitatorRejectedError) and exc.code in (
        "SIMULATION_REVERTED",
        "UNSUPPORTED_TRANSFER_METHOD",
    )
